### ISSUE TITLE
Add Support for latest create-react-app and React 18

### ISSUE THREAD

**Issue by AlexThurston** [vip-git/universal-json-schema#465]
When I create a brand new CRA:
```
npx create-react-app test
```

Install mui:
```
yarn add @mui/material @emotion/react @emotion/styled
```

Install this:
```
 yarn add react-jsonschema-form-material-ui
```

And then try to import:
```
import MaterialJsonSchemaForm from 'react-jsonschema-form-material-ui'
```

The build fails with:
```
ERROR in ./node_modules/react-jsonschema-form-material-ui/dist/bundle.js 65:9-23
Module not found: Error: Can't resolve 'tty' in '/Users/athurston/Code/portal/ls-cryptik/node_modules/react-jsonschema-form-material-ui/dist'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "tty": require.resolve("tty-browserify") }'
	- install 'tty-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "tty": false }

ERROR in ./node_modules/react-jsonschema-form-material-ui/dist/bundle.js 66:9-22
Module not found: Error: Can't resolve 'os' in '/Users/athurston/Code/portal/ls-cryptik/node_modules/react-jsonschema-form-material-ui/dist'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "os": require.resolve("os-browserify/browser") }'
	- install 'os-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "os": false }

ERROR in ./node_modules/react-jsonschema-form-material-ui/dist/bundle.js 72:9-26
Module not found: Error: Can't resolve 'stream' in '/Users/athurston/Code/portal/ls-cryptik/node_modules/react-jsonschema-form-material-ui/dist'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "stream": require.resolve("stream-browserify") }'
	- install 'stream-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "stream": false }

webpack compiled with 3 errors and 8 warnings
```

I have a feeling this has to do with either the latest version of `create-react-app` or the fact that is it using React v18.

Here are my dependencies in my `package.json`:
```
"dependencies": {
    "@emotion/react": "^11.9.3",
    "@emotion/styled": "^11.9.3",
    "@mui/material": "^5.8.6",
    "@testing-library/jest-dom": "^5.16.4",
    "@testing-library/react": "^13.3.0",
    "@testing-library/user-event": "^13.5.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-jsonschema-form-material-ui": "^3.0.0-rc.6",
    "react-scripts": "5.0.1",
    "web-vitals": "^2.1.4"
  }
```

**Comment by AlexThurston** [comment]
OK.  I've gotten a little bit further with this.  

The first thing I had to do with was react-rewired to add the missing dependencies.  With that installed and this config:
```
const { dependencies } = require("./package.json")
const webpack = require('webpack'); 
module.exports = function override(config) { 
  const fallback = config.resolve.fallback || {}; 
  Object.assign(fallback, {
    "tty": require.resolve("tty-browserify"),
    "stream": require.resolve("stream-browserify"),
    "os": require.resolve("os-browserify"),
    "buffer": require.resolve("buffer/")
  })
  config.resolve.fallback = fallback; 
  config.plugins = (config.plugins || []).concat([ 
    new webpack.ProvidePlugin({ 
      process: 'process/browser', 
      Buffer: ['buffer', 'Buffer'] 
    }) 
  ])
  return config
}
```

And removing the map generation with: `GENERATE_SOURCEMAP=false react-app-rewired start`

I was able to at least have the application build.

However, the problem now lies in the fact that in this library, react is marked as a dependency which means that it is bundled with the build.  This means that when importing you end up with two different versions of react (in my case 17 and 18) which gives this browser console err:

```
Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
    at resolveDispatcher (react.development.js:1438:1)
    at Object.useState (react.development.js:1468:1)
    at KF (bundle.js:32472:1)
    at module.exports (bundle.js:34242:1)
    at renderWithHooks (react-dom.development.js:16316:1)
    at mountIndeterminateComponent (react-dom.development.js:19986:1)
    at beginWork (react-dom.development.js:21454:1)
    at HTMLUnknownElement.callCallback (react-dom.development.js:4112:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:4161:1)
    at invokeGuardedCallback (react-dom.development.js:4223:1)
```

I think in order to fix this,`react` (and `react-dom`) need to be marked as peer dependencies in the package.json so that they are not bundled in with the release/build/etc...

**Comment by vip-git** [comment]
Hi @AlexThurston  - react 18 was not supported indeed.

I have included the support now part of `3.0.0-rc.7`.

Are you able to confirm if this resolves your issue ?

Thanks - V

**Comment by vip-git** [comment]
Thanks @AlexThurston - Appreciate your analysis there . I will close this issue for now.

Incase you find new ones - you can create a new one or re-open this one.

**Comment by nunorbatista** [comment]
I'm having the `Uncaught Error: Invalid hook call.`
My stack:

- React 18
- NextJS 12
- Material UI
- package.json: "react-jsonschema-form-material-ui": "^3.0.0-rc.9",

I downgraded to React 17 and it works.

**Comment by vip-git** [comment]
> I'm having the `Uncaught Error: Invalid hook call.`
> 
> My stack:
> 
> 
> 
> - React 18
> 
> - NextJS 12
> 
> - Material UI
> 
> - package.json: "react-jsonschema-form-material-ui": "^3.0.0-rc.9",
> 
> 
> 
> I downgraded to React 17 and it works. 

Hi @nunorbatista 

For now support for react 18 is available with '3.0.0-rc.7'

Are you able to try this. Support was dropped in future versions as it was not backwards compatible.

There will be future versions where support will be included once tests are passing for both react 17 and 18 .

In the meantime you can use '3.0.0-rc.7'

Hope that helps - thanks
